griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py

```diff
@@ -0,0 +1,133 @@
+    headers: dict = field(
+        default=Factory(lambda self: {"Authorization": f"Bearer {self.api_key}"}, takes_self=True),
+        init=False,
+    )
+
+    def __attrs_post_init__(self) -> None:
+        if self.thread_id is None:
+            self.thread_id = os.getenv("GT_CLOUD_THREAD_ID", self._get_thread_id())
+
+    @api_key.validator  # pyright: ignore[reportAttributeAccessIssue]
+    def validate_api_key(self, _: Attribute, value: Optional[str]) -> str:
+        if value is None:
+            raise ValueError(f"{self.__class__.__name__} requires an API key")
+        return value
+
+    def store(self, memory: BaseConversationMemory) -> None:
+        # serliaze the run artifacts to json strings
+        messages = [
+            {"input": json.dumps(run.input.to_dict()), "output": json.dumps(run.output.to_dict())}
+            for run in memory.runs
+        ]
+
+        # serialize the metadata to a json string
+        # remove runs because they are already stored as Messages
+        metadata = memory.to_dict()
+        del metadata["runs"]
+        metadata = json.dumps(metadata)
+
+        # patch the Thread with the new messages and metadata
+        # all old Messages are replaced with the new ones
+        response = requests.patch(
+            self._get_url(f"/threads/{self.thread_id}"),
+            json={"messages": messages, "metadata": metadata},
+            headers=self.headers,
+        )
+        response.raise_for_status()
+
+    def load(self) -> Optional[BaseConversationMemory]:
+        # get the Messages from the Thread
+        messages_response = requests.get(self._get_url(f"/threads/{self.thread_id}/messages"), headers=self.headers)
+        messages_response.raise_for_status()
+        messages_response = messages_response.json()
+        if messages_response is None:
+            raise RuntimeError(f"Error getting messages for thread {self.thread_id}")
+
+        messages = sorted(messages_response.get("messages", []), key=lambda m: m["index"])
+
+        # retrieve the Thread to get the metadata
+        thread_response = requests.get(self._get_url(f"/threads/{self.thread_id}"), headers=self.headers)
+        thread_response.raise_for_status()
+        thread_response = thread_response.json()
+
+        from griptape.memory.structure import Run
+
+        runs = [
+            Run(
+                id=m["message_id"],
+                input=BaseArtifact.from_dict(json.loads(m["input"])),
+                output=BaseArtifact.from_dict(json.loads(m["output"])),
+            )
+            for m in messages
+        ]
+        metadata = thread_response.get("metadata")
+
+        # the metadata will contain the serialized
+        # ConversationMemory object with the runs removed
+        # autoload=False to prevent recursively loading the memory
+        if metadata is not None and metadata != {}:
+            from griptape.memory.structure import BaseConversationMemory
```

Comment:
Move to top of function.

Reply:
same here, its branched

Reply:
Same solution should be applied. Move import to top of _a_ scope. Whether that's this one or a new function.